source_integration_test.go

```diff
@@ -0,0 +1,410 @@
+	is.True(errors.Is(err, sdk.ErrBackoffRetry))
+
+	_ = source.Teardown(ctx)
+}
+
+func TestSource_NonExistentTable(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	_, cfg := prepareIntegrationTest(ctx, t)
+
+	source := &Source{}
+
+	// set the table name to a unique uuid, so it doesn't exist.
+	cfg[SourceConfigTable] = "28e2d9e9-b1d6-4fd7-bfa1-183ceb3c9404"
+
+	err := source.Configure(context.Background(), cfg)
+	is.NoErr(err)
+
+	// table existence check at "Open"
+	err = source.Open(context.Background(), nil)
+	is.True(err != nil)
+	is.True(strings.Contains(err.Error(), "Cannot do operations on a non-existent table"))
+}
+
+func TestSource_CDC(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	client, cfg := prepareIntegrationTest(ctx, t)
+	//
+	testTable := cfg[SourceConfigTable]
+	source := &Source{}
+	err := source.Configure(context.Background(), cfg)
+	is.NoErr(err)
+
+	// add rows
+	err = insertRecord(ctx, client, testTable, 1, 2)
+	is.NoErr(err)
+
+	err = source.Open(context.Background(), nil)
+	is.NoErr(err)
+
+	// snapshot, one record
+	rec, err := source.Read(ctx)
+	is.NoErr(err)
+	is.Equal(rec.Payload.After, opencdc.StructuredData{PartitionKey: "pkey1", SortKey: "1"})
+
+	// add a row, will be captured by CDC
+	err = insertRecord(ctx, client, testTable, 2, 3)
+	is.NoErr(err)
+
+	// delete the latest row, will be captured by CDC
+	_, err = client.DeleteItem(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(testTable),
+		Key: map[string]types.AttributeValue{
+			PartitionKey: &types.AttributeValueMemberS{Value: "pkey2"},
+			SortKey:      &types.AttributeValueMemberN{Value: "2"},
+		},
+	})
+	is.NoErr(err)
+
+	// update the first row, will be captured by CDC
+	_, err = client.UpdateItem(ctx, &dynamodb.UpdateItemInput{
+		TableName: aws.String(testTable),
+		Key: map[string]types.AttributeValue{
+			PartitionKey: &types.AttributeValueMemberS{Value: "pkey1"}, // partition key
+			SortKey:      &types.AttributeValueMemberN{Value: "1"},     // sort key
+		},
+		UpdateExpression: aws.String("SET #attr = :newValue"),
+		ExpressionAttributeNames: map[string]string{
+			"#attr": "data", // alias 'data' to avoid reserved keyword error
+		},
+		ExpressionAttributeValues: map[string]types.AttributeValue{
+			":newValue": &types.AttributeValueMemberS{Value: "newValue"}, // the new value
+		},
+	})
+	is.NoErr(err)
+
+	// cdc
+	i := 0
+	for {
+		rec2, err := source.Read(ctx)
+		if err == nil {
+			switch i {
+			case 0:
+				is.True(rec2.Operation == opencdc.OperationCreate)
+				is.Equal(rec2.Payload.After, opencdc.StructuredData{PartitionKey: "pkey2", SortKey: "2"})
+			case 2:
+				is.True(rec2.Operation == opencdc.OperationDelete)
+				is.Equal(rec2.Payload.Before, opencdc.StructuredData{PartitionKey: "pkey2", SortKey: "2"})
+			case 1:
+				is.True(rec2.Operation == opencdc.OperationUpdate)
+				is.Equal(rec2.Payload.After, opencdc.StructuredData{"data": "newValue", PartitionKey: "pkey1", SortKey: "1"})
```

Comment:
Nitpick - can we switch cases `2` and `1`? My OCD is kicking in when I see 0 -> 2 -> 1 😄

Reply:
hahha that was a last minute change forgot about it! fixed